app/services/acuant/liveness.rb

```diff
@@ -1,22 +1,31 @@
 module Acuant
   class Liveness < AcuantBase
-    def call(liveness_body)
-      live_face_image = process_selfie(liveness_body)
-      return unless live_face_image
+    # @return [Array(Boolean, Hash)] a tuple of (overall success, error hash)
+    def call(selfie)
+      base64_selfie = Base64.strict_encode64(selfie)
+      return failure unless selfie_live?(base64_selfie)
 
       face_image_from_document = fetch_face_from_document
-      return [true, false] unless face_image_from_document
+      return failure unless face_image_from_document
 
-      [true, check_face_match(face_image_from_document, live_face_image)]
+      selfie_face_matches_document?(face_image_from_document, base64_selfie) ? success : failure
     end
 
     private
 
-    def process_selfie(liveness_body)
-      liveness_data = wrap_network_errors { liveness_service.liveness(liveness_body) }
-      return unless liveness_data
-      return unless selfie_live?(liveness_data)
-      JSON.parse(liveness_body)['Image']
+    attr_accessor :results
+
+    def failure
+      [false, results]
+    end
+
+    def success
+      [true, nil]
```

Comment:
Ok this is a nit....but uh, `nil` is not of type `Hash` so that means that the return type way up should be...

```
@return [Array(Boolean, nil), Array(Boolean, Hash)]
```

Reply:
And the reason I'm being pedantic here is because we want to make it clear what's going on so that clients of the API (which is us) are super clear about when it's OK and not OK to access an element.

And just to reiterate, I think that this would be even better as a struct not a tuple.


```
AcuantResponse = Struct.new(:success, :errors_hash)
# ...
def success
  AccuantResponse.new(true, nil)
end
```

Reply:
I ditched the nil and made it an empty hash.

Reply:
I do use the structs and they are useful in certain circumstances but I'm a bigger fan of less code (I did heavy JAVA and C++ for decades).  I'm also not a fan of comments in the code but I acquiesced here because it is a service.  If the code is unclear I need to make it that way but never at the expense of adding more code. To me the easiest code to read and maintain is the code you don't write.  Otherwise, at the extreme I'm back to JAVA again:  same functionality, more code, but less readable in my opinion, simply because there's more of it.

Reply:
I do appreciate your thorough reviews.